src/platform/networking/node/chatStream.ts

```diff
@@ -12,12 +12,81 @@ import { APIJsonData, CAPIChatMessage, ChatCompletion, rawMessageToCAPI } from '
 import { FinishedCompletion, convertToAPIJsonData } from './stream';
 
 // TODO @lramos15 - Find a better file for this, since this file is for the chat stream and should not be telemetry related
-export function sendEngineMessagesTelemetry(telemetryService: ITelemetryService, messages: CAPIChatMessage[], telemetryData: TelemetryData) {
+export function sendEngineMessagesLengthTelemetry(telemetryService: ITelemetryService, messages: CAPIChatMessage[], telemetryData: TelemetryData, logService?: ILogService) {
+	// Determine if this is input or output based on message characteristics
+	const isOutput = messages.length === 1 && messages[0].role === 'assistant';
+	const messageType = isOutput ? 'output' : 'input';
+
+	// Get the unique model call ID - it should already be set in the base telemetryData
+	const modelCallId = telemetryData.properties.modelCallId as string;
+	if (!modelCallId) {
+		// This shouldn't happen if the ID was properly generated at request start
+		logService?.warn('[TELEMETRY] modelCallId not found in telemetryData, input/output messages cannot be linked');
+		return;
+	}
+
+	// Create messages with content and tool_calls arguments replaced by length
+	const messagesWithLength = messages.map(msg => {
+		const processedMsg: any = {
+			...msg, // This preserves ALL existing fields including tool_calls, tool_call_id, copilot_references, etc.
+			content: typeof msg.content === 'string'
+				? msg.content.length
+				: Array.isArray(msg.content)
+					? msg.content.reduce((total: number, part: any) => {
+						if (typeof part === 'string') {
+							return total + part.length;
+						}
+						if (part.type === 'text') {
+							return total + (part.text?.length || 0);
+						}
+						return total;
+					}, 0)
+					: 0,
+		};
+
+		// Process tool_calls if present
+		if ('tool_calls' in msg && msg.tool_calls && Array.isArray(msg.tool_calls)) {
+			processedMsg.tool_calls = msg.tool_calls.map((toolCall: any) => ({
+				...toolCall,
+				function: toolCall.function ? {
+					...toolCall.function,
+					arguments: typeof toolCall.function.arguments === 'string'
+						? toolCall.function.arguments.length
+						: toolCall.function.arguments
+				} : toolCall.function
+			}));
+		}
+
+		return processedMsg;
+	});
+
+	// Log the messages before sending to telemetry with modelCallId as top-level property
+	const logData = {
+		modelCallId: modelCallId,
+		messageType: messageType,
+		messages: messagesWithLength
+	};
+	logService?.info(`[TELEMETRY] engine.messages.length: ${JSON.stringify(logData, null, 2)}`);
```

Comment:
I don't think we should log stuff like this to `info`.  Outgoing telemetry is logged to the telemetry channel already anyway